### PR TITLE
e2e,vnc: Use NewGuestless function to create VMIs as it can handle the larger arm64 memory requirement

### PR DESCRIPTION
**Pull request by brianmcarey**
### What this PR does

The VNC screenshot test fails on ARM64[1] as the VMI fails to start due to
not enough memory being available. The NewGuestless function that was
recently introduced into libvmi creates a guestless VMI that caters
for the increased memory requirement of ARM64[2].

[1] https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11400/pull-kubevirt-e2e-arm64/1764549347174780928
[2] https://github.com/kubevirt/kubevirt/blob/0ca03aa05ccab5fefc6bfe7b474e5e05e18929a4/tests/libvmi/factory.go#L96C1-L101C3

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

**Comment by brianmcarey** [comment]
Successful pull-kubevirt-e2e-arm64: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11366/pull-kubevirt-e2e-arm64/1762448901509484544

**Review by dhiller — APPROVED** [review]
/approve

**Comment by akalenyu** [comment]
/hold

Confused here, this is not a Cirros VM prior to this change. How come it needs more memory?
Check https://github.com/kubevirt/kubevirt/pull/11302
/cc @dankenigsberg

**Comment by brianmcarey** [comment]
> /hold
> 
> Confused here, this is not a Cirros VM prior to this change. How come it needs more memory? Check #11302 /cc @dankenigsberg

Yes its a fair point. I didn't have an environment to test this initially - I was relying on CI. 

From testing the current VMI fails to boot on the arm64 with less than 128Mi.

I can just update the ResourceMemory to 128Mi rather than changing the VMI type.

**Comment by akalenyu** [comment]
> > /hold
> > Confused here, this is not a Cirros VM prior to this change. How come it needs more memory? Check #11302 /cc @dankenigsberg
> 
> Yes its a fair point. I didn't have an environment to test this initially - I was relying on CI.
> 
> From testing the current VMI fails to boot on the arm64 cluster with less than 128Mi.
> 
> I can just update the ResourceMemory to 128Mi rather than changing the VMI type.

Hmm I would feel better if we know for sure why 128Mi is needed for this in the ARM case,
like you said, this is not a cirros VM

**Comment by zhlhahaha** [comment]
Hi, @akalenyu , When use uefi boot on aarch64 with edk2 package, qemu will create 2 pflash(64Mi each, 128Mi in total).
Additional information can be found here: 
(https://github.com/qemu/qemu/blob/bfe8020c814a30479a4241aaa78b63960655962b/hw/arm/virt.c#L126)

**Comment by dankenigsberg** [comment]
Thanks, @brianmcarey , for noticing the regression I introduced in https://github.com/kubevirt/kubevirt/pull/11340 . You are most welcome to call it out in your PR text.

As for the fix, how about 
```
func NewTiniest(opts ...Option) *kvirtv1.VirtualMachineInstance {
        opts = append(
                []Option{WithResourceMemory(qemuMinimumMemory())},
                opts...)
        return New(opts...)
}   

func qemuMinimumMemory() string{
        if isARM64() {
                // required to start qemu on ARM with UEFI firmware
                // https://github.com/kubevirt/kubevirt/pull/11366#issuecomment-1970247448
                return "128Mi"
        }
        return "1Mi"
}
```
?
 /cc @EdDev .

**Comment by brianmcarey** [comment]
> Thanks, @brianmcarey , for noticing the regression I introduced in #11302 . You are most welcome to call it out in your PR text.
> 
> As for the fix, how about
> 
> ```
> func NewTiniest(opts ...Option) *kvirtv1.VirtualMachineInstance {
>         opts = append(
>                 []Option{WithResourceMemory(qemuMinimumMemory())},
>                 opts...)
>         return New(opts...)
> }   
> 
> func qemuMinimumMemory() string{
>         if isARM64() {
>                 // required to start qemu on ARM with UEFI firmware
>                 // https://github.com/kubevirt/kubevirt/pull/11366#issuecomment-1970247448
>                 return "128Mi"
>         }
>         return "1Mi"
> }
> ```
> 
> ? /cc @EdDev .

Thanks @dankenigsberg - that looks good to me - I will update the PR here.

**Comment by EdDev** [comment]
> Thanks, @brianmcarey , for noticing the regression I introduced in #11302 . You are most welcome to call it out in your PR text.
> 
> As for the fix, how about
> 
> ```
> func NewTiniest(opts ...Option) *kvirtv1.VirtualMachineInstance {
>         opts = append(
>                 []Option{WithResourceMemory(qemuMinimumMemory())},
>                 opts...)
>         return New(opts...)
> }   
> 
> func qemuMinimumMemory() string{
>         if isARM64() {
>                 // required to start qemu on ARM with UEFI firmware
>                 // https://github.com/kubevirt/kubevirt/pull/11366#issuecomment-1970247448
>                 return "128Mi"
>         }
>         return "1Mi"
> }
> ```

`NewTiniest` sounds odd to me.
Per the description, it needs to be added to `New` as this is global.

**Comment by akalenyu** [comment]
> Hi, @akalenyu , When using UEFI boot on an AArch64 architecture with the EDK2 package, QEMU configures two pflash, each 64 MiB in size, totaling 128 MiB. The memory map for an Arm64 virtual machine can be found in the QEMU source code (https://github.com/qemu/qemu/blob/bfe8020c814a30479a4241aaa78b63960655962b/hw/arm/virt.c#L126)
> 
> The code specifies the creation of two pflash devices, with their sizes determined by the EDK2 package. The first pflash device corresponds to the UEFI firmware (/usr/share/AAVMF/AAVMF_CODE.fd), while the second is dynamically created for each VM, typically located in /var/lib/libvirt/qemu/nvram/. Each of these pflash devices is allocated 64 MiB.
> 
> This means that even thought we do not need to allocate memory for VM, we still need at least 128MiB for the uefi firmwares, otherwise the qemu process will fail.

Thanks for the explanation
/hold cancel

File: tests/libvmi/vmi.go
Comment (zhlhahaha) [inline tests/libvmi/vmi.go:275]: I would suggest to set 200Mi here. I got flakey failures when set 150Mi.

**Comment by brianmcarey** [comment]
> /test pull-kubevirt-e2e-arm64

Looks good - https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11366/pull-kubevirt-e2e-arm64/1763131740399865856

File: tests/libvmi/vmi.go
Comment (brianmcarey) [reply tests/libvmi/vmi.go:41]: It does break some instance type tests though - https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11366/pull-kubevirt-e2e-k8s-1.29-sig-compute/1763152795386515456

@EdDev - maybe it does make sense for a new factory function along the lines of `NewTiniest`?

Comment (EdDev) [reply tests/libvmi/vmi.go:41]: I do not think so, it makes little sense.
You could set whatever you need for Arm specifically, but in the end this just reflects a problem of the API.

Can you define the in-guest memory without touching the other? I think it should have an effect on the resource-memory in the end.

BTW, I think this fits now to register a default instead of setting it directly here:
https://github.com/kubevirt/kubevirt/pull/10975
But we can move it after that one gets merged.

Comment (dankenigsberg) [reply tests/libvmi/vmi.go:41]: @EdDev kubevirt API has three incompatible ways to specify VMI memory. For example if you specify the guest memory, the memory request is calculated based on it, but not if you set it explicitly to say `1Mi`.
I don't think it would be easy for `libvmi` to hide this fact. It must make it possible to keep the memory request as nil. I don't see a way around adding a `NewTiny` to be used only by tests.

File: tests/vnc_test.go
Comment (dankenigsberg) [inline tests/vnc_test.go:68]: @brianmcarey please use
```suggestion
			vmi = libvmi.NewGuestless()
```

**Review by dankenigsberg — CHANGES_REQUESTED** [review]

**Review by dankenigsberg — APPROVED** [review]

**Review by EdDev — APPROVED** [review]

**Pull request merged**
